### DISCUSSION TITLE
From event handlers to `IObservable<T>`

### DISCUSSION THREAD

**Discussion by Lehonti** [PintaProject/Pinta#904] [Ideas]
I feel that regular .NET event handlers could be awkward to work with, so what about using `IObservable<T>` instead?

I changed some code in my fork of Pinta in order to test this out, and it appears to be feasible. However, given the amount of work needed for this, I would like to discuss it first in order to know the way forward.

Take, for example, `LayerManager`. It has the following events:

```csharp
public event EventHandler? LayerAdded;
public event EventHandler? LayerRemoved;
public event EventHandler? SelectedLayerChanged;
```

...Along with the code that triggers them:

```csharp
internal void OnLayerAdded ()
{
	LayerAdded?.Invoke (this, EventArgs.Empty);
}
```

What if we started using `IObservable<T>` instead? We could know which events are available for listening by 'reading them off' the list of interfaces:

```csharp
public sealed class LayerManager :
	IObservable<LayerAdded>,
	IObservable<LayerRemoved>,
	IObservable<SelectedLayerChanged>
{ /*...*/ }
```

It would work in the following way:

```csharp
private readonly LinkedList<IObserver<LayerAdded>> layer_added_listeners = new ();

public IDisposable Subscribe (IObserver<LayerAdded> observer) // Exposed through 'IObservable<LayerAdded>' interface
{
	var node = layer_added_listeners.AddLast (observer);
	return DisposableHelper.FromDelegate (() => layer_added_listeners.Remove (node));
}

internal void OnLayerAdded ()
{
	foreach (var listener in layer_added_listeners)
		listener.OnNext (LayerAdded.Instance);
}
```

This looks more verbose, but I could encapsulate this mechanism in some sort of `EventTerminal` class, or we could use something like [Rx.NET](https://github.com/dotnet/reactive), and the code wouldn't be that much larger if it's done right.

**Comment by cameronwhite** [comment]
I'm not really clear on why this is an improvement?

Especially with the `given the amount of work needed for this` note, I think there are more worthwhile things to spend time on rather than making large changes to code that already works, unless there's a good justification for why it's needed

**Reply by Lehonti** [reply]
Sorry, I forgot to mention that this is more of a long-term plan. I agree that there are better things to do with the code at the moment.

The benefits are, in short, less coupling, greater control, and greater type safety (because listeners can choose to handle a derived message type or a base one).

I have in mind some applications where an event aggregator pattern is used and components don't even need to have a reference to the component that raised the event, but can still react to it. To take the example of `LayerManager`, a layer that is added arguably counts as an application-wide event, and one should be able to handle it anywhere in the application without having to figure out the chain of methods and constructors to be modified in order to get hold of the component raising the event, or worse, making it global (which is one of the reasons behind the `PintaCore` mess).

We can add to this the possibility of handling errors and handling events asynchronously, among other benefits related to having greater control of the behavior of event handlers beyond what a multi-cast delegate allows for.

**Reply by cameronwhite** [reply]
Thanks! Yeah, I agree that this should still be a very low priority to spend time on unless event handling becomes a big problem in the codebase
